test/persistence-hooks.suite.js

```diff
@@ -1828,6 +2068,464 @@ module.exports = function(dataSource, should) {
+                  name: 'a name',
+                  extra: undefined
+                }
+              }));
+            } else {
+              observedContexts.should.eql(aTestModelCtx({
+                data: {
+                  id: 'new-id',
+                  name: 'a name'
+                },
+                isNewInstance: true,
+                currentInstance: {
+                  id: 'new-id',
+                  name: 'a name',
+                  extra: undefined
+                }
+              }));
+            }
+            done();
+          });
+      });
+
+      it('triggers `persist` hook on replace', function(done) {
+        TestModel.observe('persist', pushContextAndNext());
+
+        TestModel.replaceOrCreate(
+          { id: existingInstance.id, name: 'replaced name' },
+          function(err, instance) {
+            if (err) return done(err);
+            
+            var expected = {
+              where: { id: existingInstance.id },
+              data: {
+                id: existingInstance.id,
+                name: 'replaced name'
+              },
+              currentInstance: {
+                id: existingInstance.id,
+                name: 'replaced name',
+                extra: undefined
+              }
+            };
+
+            var expectedContext;
+            if (dataSource.connector.replaceOrCreate) {
+              expectedContext = aTestModelCtx(expected);
+            } else {
+              expected.isNewInstance = false;
+              expectedContext = aTestModelCtx(expected);       
+          }
+
+            observedContexts.should.eql(expectedContext);
+            done();
+          });
+      });
+
+      it('triggers `loaded` hook on create', function(done) {
+        TestModel.observe('loaded', pushContextAndNext());
+
+        TestModel.replaceOrCreate(
+          { id: 'new-id', name: 'a name' },
+          function(err, instance) {
+            if (err) return done(err);
+            observedContexts.should.eql(aTestModelCtx({
+              data: {
+                id: 'new-id',
+                name: 'a name'
+              },
+              isNewInstance: true
+            }));     
+            done();
+          });
+      });
+      /* TODO: add support for option.notify = false for all methods; adding
+        support just for replaceOrCreate is not enough and make the logic 
+        inconsistent; in other words, if the non-atomic workaound is used
+        where we call `save`, `create` or `replaceById`, these methods trigger
+        hooks even though you may have `option.notify = false`, while in atomic
+        version `option.notify` will do work; so I'm proposing to create a task 
+        to change all of them together to have consistent logic and write test
+        cases for all scenarios. I also prefer to revert back this commit 
+        since this provides just partial support for `option.notify = false`
+        which is just for atomic implementation
```

Comment:
I see. My concern is about backwards compatibility. Once we publish the implementation that triggers "loaded" hook twice, then we should probably preserve that behaviour until the next major version.

I think I was not clear in my https://github.com/strongloop/loopback-datasource-juggler/pull/788#commitcomment-15763555. I'd like to see the implementation of new replace* methods modified in such way that "loaded" hook is triggered only once, regardless of which connector is in use.

I'm fine with reverting 8b202e8 as you proposed.

Shouldn't this issue be already fixed by using `replaceById` instead of `prototype.replaceAttributes` in the unoptimized path? I mean which DAO methods are triggering those two "loaded" hooks?

AFAICT, `callNonAtomicUpdateOrCreate ` calls these three methods under the hood:

1) `Model.findOne` with `options.notify:false` - no hooks should be triggered
then one of 
2a) `self.replaceById` - this triggers "loaded" hook only once
2b) `obj.save()` - this triggers "loaded" hook only once

That should trigger "loaded" only once. What am I missing?